src/main/java/dk/alexandra/fresco/lib/math/integer/binary/RepeatedRightShiftProtocolImpl.java

```diff
@@ -0,0 +1,143 @@
+
+import dk.alexandra.fresco.framework.MPCException;
+import dk.alexandra.fresco.framework.NativeProtocol;
+import dk.alexandra.fresco.framework.Protocol;
+import dk.alexandra.fresco.framework.ProtocolProducer;
+import dk.alexandra.fresco.framework.value.SInt;
+import dk.alexandra.fresco.framework.value.Value;
+import dk.alexandra.fresco.lib.field.integer.BasicNumericFactory;
+
+public class RepeatedRightShiftProtocolImpl implements RepeatedRightShiftProtocol {
+	
+	// Input
+	private SInt x;
+	private int n;
+	private SInt result;
+	
+	// Factories
+	private final BasicNumericFactory basicNumericFactory;
+	private final RightShiftFactory rightShiftFactory;
+
+	// Variables used for calculation
+	private int round = 0;
+	private SInt in, out;
+	private ProtocolProducer gp;
+	private SInt[] remainders;
+	
+	public RepeatedRightShiftProtocolImpl(SInt x, int n, SInt result,
+			BasicNumericFactory basicNumericFactory, 
+			RightShiftFactory rightShiftFactory) {
+		
+		if (n < 1) {
+			throw new MPCException("n must be positive");
+		}
+		
+		this.x = x;
+		this.n = n;
+		this.result = result;
+		
+		this.basicNumericFactory = basicNumericFactory;
+		this.rightShiftFactory = rightShiftFactory;
+	}
+
+	public RepeatedRightShiftProtocolImpl(SInt x, int n, SInt result, SInt[] remainders,
+			BasicNumericFactory basicNumericFactory, 
+			RightShiftFactory rightShiftFactory) {
+		
+		this(x, n, result, basicNumericFactory, rightShiftFactory);
+		if (remainders.length != n) {
+			throw new MPCException("Length of array for remainders must match number of performed shifts");
+		}
+		this.remainders = remainders;
+	}
+
+	public int getNextProtocols(NativeProtocol[] gates, int pos) {
+		if (gp == null) {
+
+			if (round == 0) {
+				out = basicNumericFactory.getSInt();
+
+				Protocol rightShift;
+				if (remainders != null) {
+					rightShift = rightShiftFactory.getRightShiftProtocol(x, out, remainders[round]); 
+				} else {
+					rightShift = rightShiftFactory.getRightShiftProtocol(x, out); 
+				}
+				gp = rightShift;
+			} else if (round < n - 1) {
+				in = out;
+				out = basicNumericFactory.getSInt();
+				Protocol rightShift;
+				if (remainders != null) {
+					rightShift = rightShiftFactory.getRightShiftProtocol(in, out, remainders[round]); 
+				} else {
+					rightShift = rightShiftFactory.getRightShiftProtocol(in, out); 
+				}
+				gp = rightShift;
+			} else if (round == n - 1) {
+				in = out;
+				out = null;
+				Protocol rightShift;
+				if (remainders != null) {
+					rightShift = rightShiftFactory.getRightShiftProtocol(in, result, remainders[round]); 
+				} else {
+					rightShift = rightShiftFactory.getRightShiftProtocol(in, result); 
+				}
+				gp = rightShift;
+			}
+			
+		}
+		if (gp.hasNextProtocols()) {
+			pos = gp.getNextProtocols(gates, pos);
+		} else if (!gp.hasNextProtocols()) {
+			round++;
+			gp = null;
+		}
+		return pos;
+	}
+
+	@Override
+	public Value[] getInputValues() {
```

Comment:
I know these values are null a lot of places, but we would maybe like to use them at some point, so lets try to fill them in when we can. 
This should return x I recon, and out should return res and maybe the remainders?